File: 코테/K-22[공채]/[4]양궁대회.cpp
#include <string>
#include <vector>
using namespace std;

int shotResult[11];
int shotCount[11];
int maxGap = 0;

void setShots(int left) {
    for(int i=0; i<11; ++i) shotResult[i] = 0;
    for(int i=0; i<11; ++i) {
        shotResult[i] = shotCount[i];
    }
    if(left != 0) shotResult[10] += left;
}

bool checkShots() {
    for(int i=10; i>=0; --i) {
        if(shotCount[i] == shotResult[i]) continue;
        if(shotCount[i] > shotResult[i]) return true;
        else return false;
    }
}

int getGap(vector<int> & info, int &A) {
    int sum=0;
    for(int i=0; i<info.size(); ++i) {
        if(info[i] < shotCount[i]){
            if(info[i] == 0) sum += 10 - i;
            else sum += 2*(10-i);
        }
    }
    return sum - A;
}

void DFS(int cur, vector<int> &info, int left, int &A) {
    // 마지막까지 순회하여 다 봤다면
    if(cur == 10){
        int gap = getGap(info, A);
        if(gap > maxGap || (gap ==maxGap && checkShots())) {
            maxGap = gap;
            setShots(left);
        }
        return;
    }
    // 남은 화살수가 0이라면
    if(left == 0) {
        int gap = getGap(info, A);
        if(gap > maxGap || (gap ==maxGap && checkShots())) {
            maxGap = gap;
            setShots(left);
        }
        return;
    }
    
    // 선택
    if(left >= info[cur]+1) {
        shotCount[cur] = info[cur]+1;
        DFS(cur+1, info, left - info[cur]-1, A);
        shotCount[cur] = 0;
    }
    // 안선택
    DFS(cur+1, info, left, A);
    return;
}

vector<int> solution(int n, vector<int> info) {
    vector<int> answer(1, -1);
    int A = 0;
    for(int i=0; i<info.size(); ++i)
        if(info[i] > 0) A += 10-i;
    
    DFS(0, info, n, A);
    if(maxGap > 0) {
        answer.resize(11);
        for(int i=0; i<11; ++i) {
            answer[i] = shotResult[i];
        }
    }
    
    return answer;
}

File: 코테/K-22[공채]/[4]양궁대회.js
function solution(n, info) {
  let answer = [];
  let gap = 0;

  function scoreGap(lion, apeach) {
    gap = 0;
    for (let i = 0; i < 11; ++i) {
      if (lion[i] > apeach[i]) gap += 10 - i;
      else gap -= 10 - i;
    }
    return gap;
  }

  function renewResult(lion, apeach) {
    if (gap < scoreGap(lion, apeach)) {
      gap = scoreGap(lion, apeach);
      answer = lion;
    }
  }

  function DFS(left, n, lion, apeach) {
    if (left === 0) {
      console.log(lion);
      renewResult(lion, apeach);
    }
    if (n === 11) {
      lion[10] += left;
      console.log(lion);
      renewResult(lion, apeach);
    }
    // 승리하는 경우 고려
    if (left > apeach[n] + 1) {
      lion[n] = apeach[n] + 1;
      DFS(left - (apeach[n] + 1), n + 1, lion, apeach);
      lion[n] = 0;
    }
    // 패배하는 경우 고려
    DFS(left, n + 1, lion, apeach);
  }

  const initializedScore = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
  DFS(n, 0, initializedScore, info);
  if (answer === initializedScore) answer = [-1];

  return answer;
}

solution(5, [2, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0]);
